src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,581 @@
+            _cTime = TarHelpers.GetDateTimeFromSecondsSinceEpoch(cTime);
+
+            // TODO: Read the bytes of the currently unsupported GNU fields, in case user wants to write this entry into another GNU archive, they need to be preserved.
+        }
+
+        // Reads the ustar prefix attribute.
+        // Throws if a conversion to an expected data type fails.
+        private void ReadUstarAttributes(Span<byte> buffer)
+        {
+            _prefix = TarHelpers.GetTrimmedUtf8String(buffer.Slice(FieldLocations.Prefix, FieldLengths.Prefix));
+
+            // In ustar, Prefix is used to store the *leading* path segments of
+            // Name, if the full path did not fit in the Name byte array.
+            if (!string.IsNullOrEmpty(_prefix))
+            {
+                // Prefix never has a leading separator, so we add it
+                // it should always  be a forward slash for compatibility
+                _name = string.Format(UstarPrefixFormat, _prefix, _name);
+            }
+        }
+
+        // Collects the extended attributes found in the data section of a PAX entry of type 'x' or 'g'.
+        // Throws if end of stream is reached or if an attribute is malformed.
+        private void ReadExtendedAttributesBlock(Stream archiveStream)
+        {
+            Debug.Assert(_typeFlag is TarEntryType.ExtendedAttributes or TarEntryType.GlobalExtendedAttributes);
+
+            // Regardless of the size, this entry should always have a valid dictionary object
+            _extendedAttributes ??= new Dictionary<string, string>();
+
+            if (_size == 0)
+            {
+                return;
+            }
+
+            // It is not expected that the extended attributes data section will be longer than int.MaxValue, considering
+            // 4096 is a common max path length, and also the size field is 12 bytes long, which is under int.MaxValue.
+            if (_size > int.MaxValue)
+            {
+                throw new InvalidOperationException(string.Format(SR.TarSizeFieldTooLargeForExtendedAttribute, _typeFlag.ToString()));
+            }
+
+            byte[] buffer = new byte[(int)_size];
+            if (archiveStream.Read(buffer.AsSpan()) != _size)
+            {
+                throw new EndOfStreamException();
+            }
+
+            string dataAsString = TarHelpers.GetTrimmedUtf8String(buffer);
+
+            using StringReader reader = new(dataAsString);
+
+            while (TryGetNextExtendedAttribute(reader, out string? key, out string? value))
+            {
+                _extendedAttributes ??= new Dictionary<string, string>();
+
+                if (_extendedAttributes.ContainsKey(key))
+                {
+                    throw new FormatException(string.Format(SR.TarDuplicateExtendedAttribute, _name));
+                }
+                _extendedAttributes.Add(key, value);
+            }
```

Comment:
This whole operation is super heavy on allocation and needn't be.  How commonly will this be used?

Reply:
In PAX we always use this method. And since PAX is our default format when the user does not select it, I should definitely improve this with your suggestions.